source/common/stats/grpc_metrics_service_impl.h

```diff
@@ -0,0 +1,167 @@
+
+private:
+  /**
+   * Shared state that is owned by the per-thread streamers. This allows the
+   * main streamer/TLS
+   * slot to be destroyed while the streamers hold onto the shared state.
+   */
+  struct SharedState {
+    SharedState(GrpcMetricsServiceClientFactoryPtr&& factory,
+                const LocalInfo::LocalInfo& local_info)
+        : factory_(std::move(factory)), local_info_(local_info) {}
+
+    GrpcMetricsServiceClientFactoryPtr factory_;
+    const LocalInfo::LocalInfo& local_info_;
+  };
+
+  typedef std::shared_ptr<SharedState> SharedStateSharedPtr;
+
+  struct ThreadLocalStreamer;
+
+  /**
+   * Per-thread stream state.
+   */
+  struct ThreadLocalStream
+      : public Grpc::AsyncStreamCallbacks<envoy::api::v2::StreamMetricsResponse> {
+    ThreadLocalStream(ThreadLocalStreamer& parent) : parent_(parent) {}
+
+    // Grpc::AsyncStreamCallbacks
+    void onCreateInitialMetadata(Http::HeaderMap&) override {}
+    void onReceiveInitialMetadata(Http::HeaderMapPtr&&) override {}
+    void onReceiveMessage(std::unique_ptr<envoy::api::v2::StreamMetricsResponse>&&) override {}
+    void onReceiveTrailingMetadata(Http::HeaderMapPtr&&) override {}
+    void onRemoteClose(Grpc::Status::GrpcStatus status, const std::string& message) override;
+
+    ThreadLocalStreamer& parent_;
+    Grpc::AsyncStream<envoy::api::v2::StreamMetricsMessage>* stream_{};
+  };
+
+  typedef std::shared_ptr<ThreadLocalStream> ThreadLocalStreamSharedPtr;
+
+  /**
+   * Per-thread multi-stream state.
+   */
+  struct ThreadLocalStreamer : public ThreadLocal::ThreadLocalObject {
+    ThreadLocalStreamer(const SharedStateSharedPtr& shared_state);
+    void send(envoy::api::v2::StreamMetricsMessage& message);
+
+    GrpcMetricsServiceClientPtr client_;
+    ThreadLocalStreamSharedPtr thread_local_stream_ = nullptr;
+    SharedStateSharedPtr shared_state_;
+  };
+
+  ThreadLocal::SlotPtr tls_slot_;
+};
+/**
+ * Per thread implementation of a Metric Service flusher.
+ */
+class MetricsServiceSink : public Sink {
+public:
+  // MetricsService::Sink
+  MetricsServiceSink(GrpcMetricsStreamerSharedPtr grpc_metrics_streamer);
+
+  void beginFlush() override { message_.clear_envoy_metrics(); }
+
+  void flushCounter(const Counter& counter, uint64_t) override {
+    io::prometheus::client::MetricFamily* metrics_family = message_.add_envoy_metrics();
+    metrics_family->set_name(counter.name());
+    auto* metric = metrics_family->add_metric();
+    auto* counter_metric = metric->mutable_counter();
+    counter_metric->set_value(counter.value());
+  }
+
+  void flushGauge(const Gauge& gauge, uint64_t value) override {
+    io::prometheus::client::MetricFamily* metrics_family = message_.add_envoy_metrics();
+    metrics_family->set_name(gauge.name());
+    auto* metric = metrics_family->add_metric();
+    auto* gauage_metric = metric->mutable_gauge();
+    gauage_metric->set_value(value);
+  }
+
+  void endFlush() override { grpc_metrics_streamer_->send(message_); }
```

Comment:
After the first flush, you should clear identifier info for perf reasons. Since you are reusing the same message as a member that won't happen by default (which is different from how the access log code works).